### DISCUSSION TITLE
How to work with remote GUI-based app

### DISCUSSION THREAD

**Discussion by mengineerchen** [coder/code-server#4375] [Q&A] [answered]
Hey code-server community! I am working on a customized code server project where we wanna seamlessly integrate some app into the browser IDE.
To be more precise, we have one qt designed app on the host machine. It spawns a GUI for user in local usage, kinda classic. But, we want to have it in user's remote browser.
Would you have some good practice / advice, how we can deploy it into the browser IDE provided by code-server?

**Comment by jsjoeio** [comment, marked as answer]
Hmm...good question! Maybe @bpmct knows? Would VNC work for this?

**Reply by lg-nineoneninethreenine** [reply]
> Would VNC work for this?

Actually, noVNC works for this, but you need to install it. Maybe you can offer a noVNC installation in the Docker image, since a local code-server process means, just start it via systemctl, forward the port safely, get MobaXTerm and use the built-in X11 forwarding(if you SSH). Or if you use a VM/Docker image like GKE, Oracle or DigitalOcean, you can get noVNC, make sure it's hosted at all times, get Xfce, Xorg, LightDM and set Xfce as the default desktop with `sudo update-alternatives --config x-session-manager`, choosing `/usr/bin/startxfce4` and voila! noVNC should be using Xfce or Openbox(most likely idk but possible) and you can test GUI apps via your browser.

**Reply by bpmct** [reply]
We use noVNC in Coder and it works great for this exact purpose.

**Reply by AyaanZaveri** [reply]
Is there an available Docker Image for this?

**Reply by lg-nineoneninethreenine** [reply]
I don't think so, since Coder uses system images, ie. Ubuntu/Arch Linux images with extra features loaded up, it could be futile loading them onto code-server. But installing noVNC could work. The question is, how do we install it?

**Reply by jsjoeio** [reply]
cc @bpmct

**Reply by bpmct** [reply]
You'll need to create a custom image that either

1. uses the code-server base image (`FROM: codercom/code-server:latest`), then install VNC in the image OR
2. create a custom image with VNC and code-server

**Reply by AyaanZaveri** [reply]
How would we install VNC?

**Reply by mengineerchen** [reply]
> You'll need to create a custom image that either
> 
> 
> 
> 1. uses the code-server base image (`FROM: codercom/code-server:latest`), then install VNC in the image OR
> 
> 2. create a custom image with VNC and code-server

Great! I will give it a try in my project.
